cpp/src/utilities/error_utils.hpp

```diff
@@ -0,0 +1,65 @@
+#ifndef ERRORUTILS_HPP
+#define ERRORUTILS_HPP
+
+#include <cuda.h>
+#include <cuda_runtime_api.h>
+#include <iostream>
+
+#define CUDA_TRY(call)                                                    \
+  {                                                                       \
+    cudaError_t cudaStatus = call;                                        \
+    if (cudaSuccess != cudaStatus) {                                      \
+      std::cerr << "ERROR: CUDA Runtime call " << #call << " in line "    \
+                << __LINE__ << " of file " << __FILE__ << " failed with " \
+                << cudaGetErrorString(cudaStatus) << " (" << cudaStatus   \
+                << ").\n";                                                \
+      return GDF_CUDA_ERROR;                                              \
+    }                                                                     \
+  }
+
+#define RMM_TRY(x) \
+  if ((x) != RMM_SUCCESS) return GDF_MEMORYMANAGER_ERROR;
+
+#define RMM_TRY_CUDAERROR(x) \
+  if ((x) != RMM_SUCCESS) return cudaPeekAtLastError();
+
+#define CUDA_CHECK_LAST() CUDA_TRY(cudaPeekAtLastError())
+
+#define GDF_REQUIRE(F, S) \
+  if (!(F)) return (S);
+
+namespace cudf {
+/**---------------------------------------------------------------------------*
+ * @brief Exception thrown when logical precondition is violated.
+ *
+ * This exception should not be thrown directly and is instead thrown by the
+ * CUDF_EXPECTS macro.
+ *
+ *---------------------------------------------------------------------------**/
+struct logic_error : public std::logic_error {
+  logic_error(char const* const message) : std::logic_error(message) {}
+
+  // TODO Add an error code member? This would be useful for translating an
```

Comment:
Good idea.

Reply:
Per Slack convo, decision is to hold off on implementing an `error_code` member as there is a high likelihood that the `gdf_error` codes will be re-worked and we'd like to not tie the design of this exception class to something that will change.